electrumsv/network.py

```diff
@@ -1516,77 +607,102 @@ def update_api_servers_for_wallet(self,
+            min_height = batch_heights[0]
+            max_height = batch_heights[-1]
+            esv_client = await self.get_esv_headers_client()
+            assert esv_client is not None
+            count = max_height - min_height + 1
+            logger.debug(f"Fetching headers %s from %s", count, min_height)
+            header_array = await esv_client.get_headers_by_height(min_height, count)
+            stream = BytesIO(header_array)
+            for height in batch_heights:
+                raw_header = stream.read(80)
+                try:
+                    cast(Headers, app_state.headers).connect(raw_header)
+                except MissingHeader:
+                    # NOTE(AustEcon) Recurse back until it reaches our local cache's
+                    # common parent header and successfully connects
+                    logger.error("Cannot connect missing header. Backfilling %s headers deep",
+                        MAX_CONCEIVABLE_REORG_DEPTH)
+                    from_height = max(height - MAX_CONCEIVABLE_REORG_DEPTH, 1)
 
-    def get_server_height(self) -> int:
-        main_session = self.main_session()
-        if main_session and main_session.tip:
-            return cast(int, main_session.tip.height)
-        return 0
-
-    def set_server(self, server: SVServer, auto_connect: bool) -> None:
-        config = app_state.config
-        config.set_key('server', server, True)
-        if config.get('server') is server:
-            config.set_key('auto_connect', auto_connect, False)
-            app_state.async_.spawn(self._set_main_server, server,
-                SwitchReason.user_set)
-
-    def set_proxy(self, proxy: Optional[SVProxy]) -> None:
-        if str(proxy) == str(self.proxy):
-            return
-        app_state.config.set_key("proxy", proxy, False)
-        # See if config accepted the update
-        if str(app_state.config.get('proxy')) == str(proxy):
-            self.proxy = proxy
-            logger.info("Set proxy to %s", proxy)
-            app_state.async_.spawn(self._restart_network)
-
-    def sessions_by_chain(self) -> Dict[Chain, List[SVSession]]:
-        '''Return a map {chain: sessions} for each chain being followed by any session.'''
-        result: Dict[Chain, List[SVSession]] = defaultdict(list)
-        for session in self.sessions:
-            if session.chain:
-                result[session.chain].append(session)
+                    to_height = height + 1  # And include the currently missing header
+
+                    await self._fetch_missing_headers_at_heights(
+                        [height for height in range(from_height, to_height)])
+
+                assert app_state.headers is not None
+                header = app_state.headers.lookup(double_sha256(raw_header))
+                result[height] = header
+        return result
+
+    def header_at_height(self, height: int) -> Header:
+        _header_at_height = cast(Headers, app_state.headers).header_at_height
```

Comment:
`app_state.headers` is already typed. Why you are casting it and storing a reference to this function, I am not sure.